File: Competitive Programming 3/Section 1.4/Palidromes/11221 - Magic square palindromes./Python/main.py
from  sys import stdin
import math

sqrtLength = int()
def specialPos(i):
    return sqrtLength * (i%sqrtLength) + i // sqrtLength
def magicPalindrome(s):
    for i1 in range(len(s)):
        i3,i2=len(s)-1-i1,specialPos(i1)
        i4 = specialPos(i3)
        if s[i1]!=s[i2] or s[i1]!=s[i3] or s[i1]!=s[i4]:
            return False
    return True 
def main():
    global sqrtLength
    input()
    for cs,s in enumerate(stdin,1):
        s=s.rstrip()
        s=s.translate({ord(i):None for i in ' +,.?!()'})
        sqrtLength = int(math.sqrt(len(s)))
        print("Case #{}:\n{}".format(cs,sqrtLength if sqrtLength*sqrtLength==len(s)and magicPalindrome(s) else "No magic :("))
main()

File: Competitive Programming 3/Section 1.4/Palidromes/11221 - Magic square palindromes./Java/Main.java
import java.util.Scanner;

public class Main {
    static Scanner scan = new Scanner(System.in);
    static int sqrtLength;
    static int specialPos(int i){
        return sqrtLength * (i%sqrtLength) + i / sqrtLength;
    }
    static boolean magicPalindrome(String s){
        for(int i1=0;i1<s.length();i1++){
            int i3 = s.length()-1-i1, i2 = specialPos(i1) , i4 = specialPos(i3);
            char c1 = s.charAt(i1);
            if(c1!=s.charAt(i2) || c1!=s.charAt(i3) || c1!=s.charAt(i4))
                return false;
        }
        return true;
    }
    public static void main(String[]args){
        int t = Integer.parseInt(scan.nextLine());
        for(int cs = 0 ; cs < t ; cs++){
            String s = scan.nextLine();
            s=s.replaceAll("[\\s\\+\\,\\.\\?\\!\\(\\)]","");
            sqrtLength = (int)Math.sqrt(s.length());
            System.out.printf("Case #%d:\n%s\n",cs+1,sqrtLength*sqrtLength==s.length()&&magicPalindrome(s)?sqrtLength:"No magic :(");
        }
    }
}